### PR TITLE
Fix issues with cy.type

### LINKED ISSUE #365: Chaining is broken when reusing same options object in `.type()` command
Here is a good one (Cypress 0.18.4)

This works
```js
const dontLog = {log: false}
  cy
    .visit(url)
    .get('[name=email]', dontLog).clear().type(email, dontLog)
    .get('[name=password]', dontLog).clear().type(password, {log: false})
    .get('button', {log: false}).click({log: false})
```
This does not work (enters the password in the *email* field instead
```js
const dontLog = {log: false}
  cy
    .visit(url)
    .get('[name=email]', dontLog).clear().type(email, dontLog)
    .get('[name=password]', dontLog).clear().type(password, dontLog)
    .get('button', {log: false}).click({log: false})
```

The above code is inside custom command `Cypress.addParentCommand('login') {...}`

I am attaching the video (without clicking submit so it stops and shows the problem)

login-problem-gif

### LINKED ISSUE #420: `cy.clear()` not behaving the same as `{selectall}{backspace}`
Using the library [jquery-maskmoney](http://plentz.github.io/jquery-maskmoney/) (v3.0.0) and testing typing into an input, the `cy.clear()` command does not behave the same as using `cy.type({selectall}{backspace})`.

```javascript
it('clear and type money value', function(){
    cy
        .visit('http://plentz.github.io/jquery-maskmoney/')
        .get("#demo8").clear().type('1000')
            .invoke('val').should('eq', '10.00')  // this fails
})

it('selectall and backspace money value', function(){
    cy
       .visit('http://plentz.github.io/jquery-maskmoney/')
       .get("#demo8").type('{selectall}{backspace}1000')
           .invoke('val').should('eq', '10.00')  // this passes
})
```

screen shot 2017-02-09 at 3 25 19 pm

### LINKED ISSUE #586: Update/improve cy.type validation logic
It appears fairly common that users need to type into an input that has an invalid *type* attribute (`type="STRING"`, `type="username"`), triggering the error `cy.type() can only be called on textarea or :text`. 

Here are two possible ways to improve the experience for this use case:

1. Allow validation to be bypassed with either `{ force: true}` or `{ validation: false }`. Also, improve the error message so it's clear what `:text` means and include details on how to bypass validation.

2. Instead of whitelisting input types, blacklist certain types instead (e.g. checkbox). This would provide a good error message if typing into something that's blacklisted, but allow users to type into inputs with types that aren't accounted for, just as a browser actually allows.

### LINKED ISSUE #593: if chaining .clear() with .type(), do not issue 'blur', 'click', etc to input in between
- Operating System: Mac
- Cypress Version: 0.19.4
- Browser Version: Chrome

### Is this a Feature or Bug?
Enhancement

### Current behavior:
If I write a test:

```js
cy.get('input').clear().type('Jane')
```

After the `.clear()`, Cypress blurs off the input, then attempts to click back onto the input at the beginning of the `type`. 

These events firing can cause some unexpected results when I'm essentially expecting the caret to clear the input and stay in the input to then type. 

### Desired behavior:

If `.clear()` is followed with `.type()`, keep the caret in the input and do not fire the events to blur the input after clear and click on the input at the start of type.

### LINKED ISSUE #596: Cannot type in draftjs-based editors
- Operating System:
Ubuntu 16.04
- Cypress Version:
Cypress CLI: 0.13.1
Cypress App: 0.19.4
- Browser Version:
Chrome 59
Electron 53

### Is this a Feature or Bug?
Bug

### Current behavior:
image

### Desired behavior:
image

### How to reproduce:
https://github.com/dziamid/cypress-draftjs

#### Test code:

```js
describe('React RTE', function(){

  it('i cannot simply type', function () {

    cy
        .visit('https://react-rte.org/demo')
        .get('[contenteditable]').eq(0).type('I am typing')
        .get('textarea').eq(0).contains('I am typing')
  });


})
```

### LINKED ISSUE #610: applies blur to wrong element after a series of tests
### Current behavior:
when doing tests on input validation (clearing, typing, blurring), for some reason its giving me "failed because this element you are chaining off of has become detached or removed from the DOM:" error unexpectedly, if i comment out some of the previous tests run it works.

and it shows the blur being applied to the last input of the previous test for some reason

### Desired behavior:
blur should be applied to current element

### How to reproduce:
look at code below, if i comment out the first password complexity tests then the email passes. but i cannot get both to run and pass.

#### Test code:

```js
const blurOptions = {
  force: true
};

const invalidSelector = 'hwui-FormInput__has-errors';
const invalidCheckboxSelector = 'hwui-Checkbox__has-errors';

describe('validates fields', () => {
  beforeEach(() => {
    cy.visit(cy.urls.register);
  });

  it('firstName', () => {
    cy.get('#id_first_name').as('firstName');
    cy.get('@firstName').focus().blur();
    cy.get('@firstName').parent().should('have.class', invalidSelector);
    cy.get('@firstName').type('asd').blur(blurOptions);
    cy.get('@firstName').parent().should('not.have.class', invalidSelector);
  });

  it('lastName', () => {
    cy.get('#id_first_name').as('lastName');
    cy.get('@lastName').focus().blur();
    cy.get('@lastName').parent().should('have.class', invalidSelector);
    cy.get('@lastName').type('asd').blur(blurOptions);
    cy.get('@lastName').parent().should('not.have.class', invalidSelector);
  });
});
```

### Additional Info (images, stack traces, etc)
note in the picture i am selecting the email field but in the console it is saying that the password field is removed. it also happens even if both test are on the same input

screen shot 2017-08-28 at 2 40 00 pm

- Operating System: macOS Sierra
- Cypress Version: 
Cypress CLI: 0.13.1
Cypress App: 0.19.4

### LINKED ISSUE #651: Trigger "change" event correctly
Cypress 0.19.4 / 0.20.0 does not generate "change" event same way as the browser. For example in vanilla JS TodoMVC this leads to different results

```js
// fires change event
cy.find('input').type('foo{enter}')
// does not fire change event
cy.find('input').type('foo').type('{enter}')
```
which shows up in this example repo https://github.com/bahmutov/add-todo-test/issues/1

Good url to test http://todomvc.com/examples/vanillajs/

test
```js
const getItems = () =>
  cy.get('.todo-list li')

it('adds 1 todo', () => {
  const url = 'http://todomvc.com/examples/vanillajs/'
  cy.visit(url)
  getItems()
    .should('have.length', 0)
  cy.get('.new-todo').type('pass this test').type('{enter}')
  getItems()
    .should('have.length', 1)
})
```

### LINKED ISSUE #940: Digits Scrambled when using type on tel input field with input mask
- Operating System: Windows 10
- Cypress Version: 1.1.0
- Browser Version: Chrome

### Is this a Feature or Bug? 
Bug

### Current behavior: 
When using .type() on tel inputs with masks, the numbers are entered out of order.  That is, digits are added to the end of the field when they meet a special character.

### Desired behavior:
The digits are entered into the field in order.

### How to reproduce:
Call .type() on a tel input field with an input mask that provides hyphens, parentheses, or other formatting

#### Test code:

```js
cy.get("#PrimaryPhone").type("1234567890").should('have.value', "(123) 456-7890")
//The above expected output is based on the input mask used by my application.  
//Your mileage may vary.
```

### Additional Info (images, stack traces, etc)

### LINKED ISSUE #1002: .clear() on an input at maxlength doesn’t work
- Operating System: 10.12.6
- Cypress Version: 1.1.3
- Browser Version: n/A

### Is this a Feature or Bug?

A bug.

### Current behavior:

Calling `.clear()` on an input with a value as long as its `maxlength` attribute value leaves the value untouched.

### Desired behavior:

Calling `.clear()` on an input with a value as long as its `maxlength` attribute value empties the field.

### How to reproduce:

1. Create an input.
2. Set its `maxlength` to `n`.
3. Set its `value` to a string of `n` characters.
4. Call `.clear()` on it.

#### Test code:

```js
cy.get('input[maxlength]').clear()
```

### Additional Info (images, stack traces, etc)

I suspect the problem to be here:

https://github.com/cypress-io/cypress/blob/324d549497345246ca76dbfb6d9ede26bb142048/packages/driver/src/cypress/keyboard.coffee#L541-L548

There is a max length, but the value is already at its max length, so it results on a no-op, when it really should clear the field.

### LINKED ISSUE #1108: Using Contenteditable and type - caret does not move
- Operating System: Mac OS Sierra
- Cypress Version: 1.4.0
- Browser Version: Chrome 63.0.3239.108

### Is this a Feature or Bug?
Bug

### Current behavior:
Using the Cypress `type` function on a `contenteditable` input doesn't move the caret along with the text.  At first I thought this might be specific to React, but actually it does the same thing with a plain `contenteditable` input on any HTML page.

### Desired behavior:
To move the caret with the text so you can simulate typing and do things like `{enter}`

### How to reproduce:
1. Create a `contenteditable` html element
2. Use `type()` on that element

#### Test code:

```
<p contenteditable="true">
    Hello
</p>

cy.get('p').type('hello', {delay: 1000})
```

The delay is helpful so you can actually see the caret as it's being typed out.

### Additional Info (images, stack traces, etc)
contenteditable

### LINKED ISSUE #1171: Change event is not fired on number input
- Operating System: MacOs
- Cypress Version: 1.4.1
- Browser Version: Chrome 63 & Electron 53
- React: 16.2.0

### Is this a Feature or Bug?
Bug

### Current behavior:
Change event is not fired on "number" type inputs after .type() nor .click() or .focus() to other element.
It works fine with "text" input types

### Desired behavior:
Change event should fire with number input just like with text input type

### How to reproduce:
https://stackblitz.com/edit/react-byhe3e?file=index.js

#### App code:

```js
import React, { Component } from 'react';
import { render } from 'react-dom';

class App extends Component {
  state = {
    inputTextValue: '',
    inputNumberValue: ''
  };

  handleTextChange = (event) => {
    this.setState({ 
      inputTextValue: event.currentTarget.value,
    });
  }

 handleNumberChange = (event) => {
    this.setState({ 
      inputNumberValue: event.currentTarget.value,
    });
  }

  render() {
    return (
      <div>
        <div>
          Input text: {this.state.inputTextValue} <br/>
          <input id="textInput" onChange={this.handleTextChange } value={this.state.inputTextValue} type="text" /> <br/><br/>
        </div>
        <div>
          Input number: {this.state.inputNumberValue} <br/>
          <input id="numberInput" onChange={this.handleNumberChange } value={this.state.inputNumberValue} type="number" />
        </div>
      </div>
    );
  }
}

render(<App />, document.getElementById('root'));
```
**Test code:**
```js
describe("Change event bug repro", function() {
    it("Should fire change on textInput", function() {
        cy.visit("https://react-byhe3e.stackblitz.io");

        cy
            .get("#textInput")
            .type("hello")
            .should("have.value", "hello");
        cy.get("#inputTextValue").contains("hello");
    });

    it("Should fire change on numberInput entering integers", function() {
        cy
            .get("#numberInput")
            .type("1234.56")
            .should("have.value", "1234");
        cy.get("#inputNumberValue").contains("1234.56");
    });

    it("Should fire change on numberInput entering decimals", function() {
        cy
            .get("#numberInput")
            .type("1234.56")
            .should("have.value", "1234.56");
        cy.get("#inputNumberValue").contains("1234.56");
    });
});
```
### Additional Info (images, stack traces, etc):
related issue: https://github.com/cypress-io/cypress/issues/816
test results:
image

### LINKED ISSUE #1209: search-input.sync method not triggered when using the  cy.get(' ').type('') command
### Current behavior:
  I am using cypress to test a Vuetify component V-select.
  After  `cy.get(' ').type('')`  command, the     `:search-input.sync="search"`      method of vue components not triggered.

### Desired behavior:

The search method should be triggered after I type some something into the search-input

### How to reproduce:
```html
<v-select
  label="Async items"
  autocomplete
  :loading="loading"
  dark
  multiple
  cache-items
  chips
  required
  :items="items"
  :rules="[() => select.length > 0 || 'You must choose at least one']"
  :search-input.sync="search"
  v-model="select"
></v-select>
```

### LINKED ISSUE #1366: Cannot type into a datetime-local input field
- Operating System: Win 10
- Cypress Version: 2.0.3
- Browser Version: Chrome 64

### Is this a Feature or Bug?
Bug

### Current behavior:
The input field doesn't update with a value when you type in text

### Desired behavior:
The value should appear in the input

### How to reproduce:
<input type="datetime-local"></input>

#### Test code:

Typescript:
```cy.get("input[type='datetime-local']").type("1");```

### Additional Info (images, stack traces, etc)

I can see from the log it correctly "yields" the input element. It just doesn't visually display anything in the input box (when the type command executes).

### LINKED ISSUE #1381: .type() does not type random characters (reproducibility is 10%)
- Operating System: Windows 10
- Cypress Version: 1.4.2
- Browser Version: Chrome 63

### Current behavior:
```js
cy.get('some css locator').first().clear().type('1000');
```
the field I'm typing to is input with `type='number'`
10% times I get input `100` (without the last character)

### LINKED ISSUE #1684: type() won't tipe in autoformating field
I'm testing a client's react app and it has a credit card page. type() is not filling the fields, which are designed to autoformat the input to match a credit card or date format. (4242424242424242 -> 4242 4242 4242 4242).

Is there a way to deal with this behavior or to simply change the field values directly?

### LINKED ISSUE #1686: cy.type() on a DIV
I have an app which is a text editor and the DOM is formatted using regular elements (h1, div, span) and they become actionable when clicked and the user can type as in a regular textarea  or input . Cypress won't let me type on it, though. 

I tried using force, with no success.

Is there any way to make it work or I'm stuck with Dusk? :(

EDIT:

I tried the following solution but it only types the first character on the desired field and the rest switches to another one.

Here's my DOM and how it should work 
https://imgur.com/fCjZLRK

Here's how it's happening with Cypress
https://imgur.com/Oynzlrd

And here's my current code

```js
cy.get('.editor-title').find('[contenteditable]').click().type('An unexpected party', {
      force: true,
    });
```

### LINKED ISSUE #1926: Cypress creates a new line (<br> tag) after entering the first letter of a word in text area.
### Current behavior:
Cypress creates a new line after entering the first letter of a word.
screen shot 2018-06-11 at 4 53 27 pm

### Desired behavior:
The code should not create new line and enter letter like "Who" 
not like "W
ho"

### Steps to reproduce:

Code:-
```js
cy.get("<text area locator>").should('be.visible')
        .type("Who", { force: true })
```

The above code is creating a new line after entering 'W' letter.
```
W
ho
```

### Versions
Cypress: 3.1.0
Operating System: MacOS Sierra v10.12.6 (16G1212)
Browser: Chrome v66.0.3359.181 (Official Build) (64-bit)

### LINKED ISSUE #2056: Problem with clear() and type() on text input field (maxlength)
### Current behavior:
Cypress is not able to clear() and type() in text input field.
The problem is not constant, it appears in my tests only in few scenarios.

### Desired behavior:
The clear() and type() actions should always work for input text type field.
### Steps to reproduce:
My test looks like follow:
I am updating existing element. The update form input text type field:
```html
\<input _ngcontent-trw-171="" class="form-control ng-untouched ng-pristine ng-valid" id="coalTypeDescr" maxlength="25" name="coalTypeDescr" ngcontrol="coalTypeDescr" required="" type="text"\>
```

My test first needs to clear existing value and add new one:
```js
   cy.get('#coalTypeDescr')
      .clear()
      .type("CoalToEditEdited")
```

cypressinputfieldproblemwithclear
cypressinputfieldproblemwithtype

Console doesn't show any problem with both functions.

### Versions
Cypress 3.0.1, macOS 10.13.5, Chrome

### LINKED ISSUE #2096: Froala (WYSIWYG HTML Editor) does not read correctly the virtual keyboard input of Cypress
### Current behavior:
The virtual keyboard inputs does not work as expected, especially regarding the meta keys. In the specific case of the ENTER key, the expected line feed doesn't seem to be outputted at all on the Froala editor.
Considering the example stringto be outputted by Cypress in the Froala Editor:
```
Hello{enter}{enter}{enter}world!
``` 
Then the result is actually looking like this:
image

### Desired behavior:
When running my end-to-end tests with Cypress on a webpage including a Froala editor, the virtual keyboard inputs should behave as a real.
If the test run by Cypress inputs the following string (same as above) in the Froala editor:
```
Hello{enter}{enter}{enter}world!
``` 
I intend to get 3 line feeds between `Hello` and `world!`, producing an inner HTML looking like this:
```
<p>Hello</p><p><br></p><p><br></p><p>world!</p>
```
The result should look like following in an actual webpage:
image

### Steps to reproduce:

I made a simple GitHub repository that repoduces the problem live.
1) Clone `https://github.com/jack-in-the-box/froala-vs-cypress.git` with git
2) At the root of the produced folder, run `npm run gui` (this installs the dependencies of Cypress and then, runs its GUI)
3) Click on the `froala.spec.js` link:
image
4) Cypress runs the test on Chrome 67 for you
5) The actual routine of the test can be seen on the file `cypress/integration/froala.spec.js`

### Versions

Latest version of Cypress, on Ubuntu, running Google Chrome 67

### LINKED ISSUE #2110: Number input doesn't increment on .type({uparrow})
### Current behavior:

`.type({uparrow})` doesn't increment the number input.
Additionally I cannot `invoke('stepUp')` either.

### Desired behavior:

Increments the number input's value on `.type({uparrow})` and `invoke('stepUp')`.

### Steps to reproduce:

https://github.com/hally9k/cypress-test-tiny
Clone this fork, `npm install` and `npm start`

### Versions

macOS 10.13.5
Cypress 3.0.2

### LINKED ISSUE #2173: .type() negative number into input not working
### Current behavior:

When typing negative number into input **with type "number" set**, minus is ignored and possitive number is typed instead.

used method: `cy.type()`

numbertypebug

### Desired behavior:

Type negative number.

### Steps to reproduce:

html

### Versions

Cypress: 3.0.2
Windows 10
Chrome 67

### LINKED ISSUE #2187: {uparrow} and {downarrow} do not move cursor in textareas and contenteditables
Currently we only send the events, we don't actually move the cursor

### PR DESCRIPTION
**Pull request by Bkucera**
this grew to a large PR fixing many cy.type issues.

fix #365
fix #420
fix #586 
fix #593 
fix #596 
fix #610 
fix #651
fix #940
fix #1002 
fix #1108
fix #1171
fix #1209 
fix #1234 
fix #1366
fix #1381 
fix #1684 
fix #1686
fix #1926 
fix #2056
fix #2096 
fix #2110 
fix #2173
fix #2187

### CODE REVIEW

File: packages/driver/src/cy/commands/actions/click.coffee [hunk 1]
```diff
@@ -199,6 +203,11 @@ module.exports = (Commands, Cypress, cy, state, config) ->
             if domEvents.mouseDown.preventedDefault or not $dom.isAttached($elToClick)
               afterMouseDown($elToClick, coords)
             else
+              if $elements.isInput(el) or $elements.isTextarea(el) or $elements.isContentEditable(el)
+                if $elements.isNeedSingleValueChangeInputElement(el)
```
Comment (jennifer-shehane) [inline packages/driver/src/cy/commands/actions/click.coffee]: cannot tell what this means from its name 🤔

Comment (jennifer-shehane) [reply packages/driver/src/cy/commands/actions/click.coffee]: oh lord, read the comments - maybe...this...is...fine... 😳

**Comment by sophielenoir** [comment]
Hi, are we sure that all the listed issues are fixed ? Indeed, I see that they are supposed to be fixed since the 3.0.3 version but I am using the 3.1 version and still experiencing issues with the "type" function as I attempt to call it in a froala editor. 
I get the following message : 
image

**Comment by phoet** [comment]
i might have found another issue when usinig `min` and `max` fields on an `input[type=number]`. since i'm in the middle of a refactoring and upgrading cypress did not help, i'm currently working around this issue with a custom command:

```
Cypress.Commands.add('replace', { prevSubject: true }, (subject, value) => {
  // only relyable way to replace the content without scrambling the value (adding min or max)
  cy.wrap(subject).clear({ force: true });
  cy.wrap(subject).clear({ force: true }).type(value);
});
```

**Comment by Bkucera** [comment]
@phoet could you open an issue or add your comment to one of the related issues? merged PRs are not very visible. Thanks

**Pull request merged**